### ISSUE TITLE
check that sbt is on your PATH and that your config is compatible

### ISSUE THREAD

**Issue by neverov** [ensime/ensime-server#670]
**TL;DR**
- check that `sbt` really is on your `PATH`
- check the variable `ensime-sbt-command` is non-nil (and is valid)
- enable `(setq debug-on-error t)`, try again and open up a ticket with the backtrace (unless the problem becomes obvious), a copy of your `.ensime` (or link to the project), and your operating system details.

---

I'm trying to use `ensime-sbt` plugin, which is defined in `.sbt/0.13/plugins/plugins.sbt` like this: 

```
resolvers += Resolver.sonatypeRepo("snapshots")

addSbtPlugin("org.ensime" % "ensime-sbt" % "0.1.5-SNAPSHOT")
```

I'm generating `.ensime` config using `gen-ensime`. It generates successfully, but when I try to load it in emacs I get the following error: `check that sbt is on your PATH and that your config is compatible with https://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]`. The generated `.ensime` config file is [here](https://github.com/neverov/scala-ensime-test/blob/master/.ensime).

I use GNU Emacs 24.3.1 (x86_64-apple-darwin14.0.0, NS apple-appkit-1343.13).
I created the following tiny project to illustrate the issue: https://github.com/neverov/scala-ensime-test.
My emacs configuration is defined here: https://github.com/neverov/emacs-config (it's located in `~/.emacs.d/` on my machine).

**Comment by xyalan** [comment]
I have the same issue in gui mode.In terminal mode work well.

**Comment by fommil** [comment]
solutions at the top

**Comment by peterbecich** [comment]
I had the same error, and eventually found a different solution.

After updating the Ensime package in Emacs to 20150211.1505, that error appeared.  Deleting `.ensime`, `.ensime_cache`, the Ivy caches in `~/.ivy2` did not fix it.  SBT was on the Path, although the variable `ensime-sbt-command` was nil.  Changing the SBT version used by the project from 0.13.5 to 0.13.7 and regenerating the `.ensime` file replaced the error with 
`could not find or load main class org.ensime.server.Server`.

This went away by just deleting the Ensime package in `*Packages*` and then installing it again.  Now it works again.

**Comment by hedefalk** [comment]
I get this same error. I'm just trying to start up Emacs as a reference when working with Atom. I'm in no way an Emacs users so I'm really pretty lost here.

I was using an older Ensime with shell-script for starting the server but as I understand it, it's to be started with sbt now so I need to check how that is done.

I have generated my .ensime with ensime-sbt 0.1.5 and it seems to be right.

My ~/emacs.d/init.el looks like:
```
;; if you're new to the MELPA package manager, just include
;; this entire snippet in your `~/.emacs` file and follow
;; the instructions in the comments.
(require 'package)
(add-to-list 'package-archives
         '("melpa" . "http://melpa.milkbox.net/packages/") t)
(package-initialize)

(when (not package-archive-contents)
  (package-refresh-contents))

;; Restart emacs and do `M-x package-install [RETURN] ensime [RETURN]`
;; To keep up-to-date, do `M-x list-packages [RETURN] U [RETURN] x [RETURN]`

;; If necessary, make sure "sbt" and "scala" are in the PATH environment
(setenv "PATH" (concat "/usr/local/bin/sbt:" (getenv "PATH")))
(setenv "PATH" (concat "/usr/local/bin/scala:" (getenv "PATH")))


(require 'ensime)
(add-hook 'scala-mode-hook 'ensime-scala-mode-hook)
```

Starting ensime I get:

```
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/build.sbt
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/plugins.sbt
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/build.properties
condition-case: check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]
```

The steps at the top, can I pretty please get "monkey instructions" for performing these steps? I'm really no Emacs user :)

**Comment by fommil** [comment]
hmm. go into `/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I` and type `sbt update`. that might shed some light on this.

**Comment by fommil** [comment]
re: "monkey instructions" just copy and paste

```
(setq debug-on-error t)
```

into your .emacs and restart emacs and try again.

**Comment by hedefalk** [comment]
Thanks for quick reply, @fommil !

sbt:
```
var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I $ sbt update
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
[info] Loading global plugins from /Users/viktor/Dropbox/Shared settings/sbt/plugins
[info] Updating {file:/Users/viktor/.sbt/0.13/plugins/}global-plugins...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Loading project definition from /private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project
[info] Updating {file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/}ensime_update_17076b0i-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Set current project to ensime_update_17076b0i (in build file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/)
[info] Updating {file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/}ensime_update_17076b0i...
[info] Resolving jline#jline;2.12 ...
[info] Done updating.
[success] Total time: 6 s, completed Apr 2, 2015 10:48:24 PM
```

**Comment by hedefalk** [comment]
emacs:

```
Debugger entered--Lisp error: (error "check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]")
  signal(error ("check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]"))
  error("check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]")
  (condition-case ex (if ensime-auto-generate-config (ensime--maybe-refresh-config nil (quote ensime--maybe-update-and-start) (quote (lambda (reason) (ensime--maybe-update-and-start)))) (ensime--maybe-update-and-start)) ((quote error) (error (format "check that sbt is on your PATH and that your config is compatible with %s [%s]" "http://github.com/ensime/ensime-server/wiki/Example-Configuration-File" ex))))
  ensime()
  call-interactively(ensime record nil)
  command-execute(ensime record)
  execute-extended-command(nil "ensime")
  call-interactively(execute-extended-command nil nil)
```

**Comment by fommil** [comment]
in the buffer called `*scratch*` type

```
(message ensime-sbt-command)
```

and the go to the end of the line and type `C-x C-e` (which executes the lisp its on) and see what gets printed in the minibuffer at the bottom.

**Comment by fommil** [comment]
btw, are you definitely using paulp's sbt (and not some cheap imitation)?

**Comment by hedefalk** [comment]
So (message ensime-sbt-command) gives me 'nil', but that's not really news, right? Question is why, no?

Do I need to use paulp's sbt? Did before but it broke somehow…

**Comment by fommil** [comment]
try removing the `PATH` lines. I'm pretty sure /usr/local is on the default path anyway.

**Comment by hedefalk** [comment]
I actually added them just before posting, had the same error before. Get the same stack too.

**Comment by fommil** [comment]
doh! we're both eejjits. Of course, PATH is supposed to be the directory not the files themselves. just add /usr/local/bin not the scripts.

**Comment by hedefalk** [comment]
Dooh! It's late here in Sweden, I'll blame the clock.

Still the same problem though. Same stack.

Both sbt and Scala are installed via homebrew on Mac. I just "which sbt" and copied it to my init.el. Homebrew uses softlinks, that can't be the problem though can it?

```
$ ll `which sbt`
lrwxr-xr-x  1 viktor  admin    28B Jun 18  2014 /usr/local/bin/sbt@ -> ../Cellar/sbt/0.13.5/bin/sbt


/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I $ cat /usr/local/Cellar/sbt/0.13.5/bin/sbt
#!/bin/sh
test -f ~/.sbtconfig && . ~/.sbtconfig
exec java -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled ${SBT_OPTS} -jar /usr/local/Cellar/sbt/0.13.5/libexec/sbt-launch.jar "$@"
```

**Comment by fommil** [comment]
use https://github.com/paulp/sbt-extras/blob/master/sbt

**Comment by hedefalk** [comment]
```
;; (setenv "PATH" (concat "/usr/local/bin:" (getenv "PATH")))
(setq exec-path (append exec-path '("/usr/local/bin")))
```
PATH env doesn't seem to work. Googled a bit and found http://emacswiki.org/emacs/ExecPath so I just tried it with the exec-path list and it now works.

```
(message exec-path)

eval: Wrong type argument: stringp, ("/usr/bin" "/bin" "/usr/sbin" "/sbin" "/Applications/Emacs.app/Contents/MacOS/libexec" "/Applications/Emacs.app/Contents/MacOS/bin" "/usr/local/bin")

(message (getenv "PATH"))
"/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"

**Comment by hedefalk** [comment]
Only difference I see is that it we prepended to the env PATH string and appended on the exec-path list. They're definitely not in sync though and only exec-path seems to bite for me.

**Comment by fommil** [comment]
interesting. Feel free to update the wiki.

**Comment by fommil** [comment]
heh, I just realised you've actually commited to paulp's sbt script, so I'm guessing you knew about it already :smile:

**Comment by hedefalk** [comment]
@fommil, yeah, I used to use it, but at some point it just stopped working. Sbt started and then quit again with no error messages. I couldn't figure it out and at the moment I wasn't using anything of the extra stuff anyways so I just started using vanilla sbt instead. I might give it another shot!

**Comment by mateialexandru** [comment]
Thanks guys!

**Comment by rnjn** [comment]
for the record - 
I had to set :java-home explicitly (marked as optional in the docs(https://github.com/ensime/ensime-server/wiki/Example-Configuration-File) on my mac to fix this error.

**Comment by fommil** [comment]
@rnjn where does it say optional? If it says that, it's an error and we should fix the docs

**Comment by rnjn** [comment]
@fommil my bad. should have read the docs more carefully. the docs do say :java-home is required.

**Comment by ajknoll** [comment]
I had this problem, and discovered that `scala-version` was unset in `.ensime` as generated by `sbt "ensime generate"`.

**Comment by fommil** [comment]
You're using an ancient version of the sbt plugin

**Comment by prassee** [comment]
Is this solved .. I still get this with sbt 0.13.7 in MacOS
